File: src/main/java/KNN.java
package com.alwin;

import java.io.File; // TODO remove unneeded imports
import java.io.FileWriter;
import java.io.FileNotFoundException;
import java.io.IOException;
import java.io.BufferedWriter;
import java.io.PrintWriter;
import java.util.Scanner;

import java.util.Arrays;
import java.util.Iterator;

import org.apache.spark.SparkConf;
import org.apache.spark.SparkContext;
import org.apache.spark.broadcast.Broadcast;
import org.apache.spark.api.java.JavaPairRDD;
import org.apache.spark.api.java.JavaRDD;
// import org.apache.spark.api.java.JavaSparkContext;
import org.apache.spark.api.java.function.FlatMapFunction;
import org.apache.spark.api.java.function.Function2;
import org.apache.spark.api.java.function.PairFunction;
import org.apache.spark.mllib.regression.LabeledPoint;
import org.apache.spark.mllib.util.MLUtils;

import scala.Tuple2;

public class KNN {

	public static void main(String[] args) throws Exception {
		if (args.length != 4) { // TODO add check that instances of train and test are same length for each data point
			System.out.printf("Usage: KNN <input dir that holds train data> <output dir> <test data file> <k>\n");
			System.exit(-1);
		}
		  
		long startTime = System.nanoTime();

		Integer k = Integer.valueOf(args[3]); // k for kNN

		SparkConf conf = new SparkConf().setAppName("KNN Spark");
		SparkContext sc = new SparkContext(conf);

		// final Broadcast<Integer> broadcastK = sc.broadcast(k);
		
		// Load and parse the data files.
		JavaRDD<LabeledPoint> trainData = MLUtils.loadLibSVMFile(sc, args[0]).toJavaRDD(); // TODO make tuplle with 1sst part being index? // zipWithIndex
		JavaPairRDD<Long, LabeledPoint> indexedTrainData = trainData
			.zipWithIndex()
			.map(new Function<Long, LabeledPoint, Long, LabeledPoint>() {
				// public Iterator<String> call(String s) { return Arrays.stream(s.split(" ")).iterator(); }
				// @Override
				public Tuple2<Long, LabeledPoint> call(LabeledPoint point, Long index) {
					return new Tuple2<Long, LabeledPoint>(index, point);
				}
		}); // TODO do this to test and/or train?
		//val indexKey = withIndex.map{case (k,v) => (v,k)}  //((0,a),(1,b),(2,c))
		//Note that some RDDs, such as those returned by groupBy(), do not guarantee order of elements in a partition. The index assigned to each element is therefore not guaranteed, and may even change if the RDD is reevaluated. If a fixed ordering is required to guarantee the same index assignments, you should sort the RDD with sortByKey() or save it to a file.
		JavaRDD<LabeledPoint> testData = MLUtils.loadLibSVMFile(sc, args[2]).toJavaRDD(); // loads sparse vector

		trainData.cache();
		// int trainingFrequency = trainData.groupBy("label").count();
		// System.out.println("trainingFrequency: " + trainingFrequency);

		// System.out.println(trainData.rdd);
		System.out.println("HERE training data rdd: " + trainData.toString());

		// collect RDD for printing
        for(LabeledPoint line:trainData.collect()){
            System.out.println("* "+line.toString());
		}


		// collect RDD for printing
        for(LabeledPoint line:indexedTrainData.collect()){
            System.out.println("* "+line.toString());
		}
// * (0.0,(7,[0,1,2,3,4,5,6],[0.0,0.0,0.0,0.0,0.0,0.0,0.0]))
// * (2.0,(7,[0,1,2,3,4,5,6],[2.0,2.0,2.0,2.0,2.0,2.0,2.0]))
// * (2.0,(7,[0,1,2,3,4,5,6],[1.0,1.0,1.0,1.0,1.0,1.0,1.0]))
// * (1.0,(7,[0,1,2,3,4,5,6],[10.0,10.0,10.0,10.0,10.0,10.0,10.0]))
// * (5.0,(7,[0,1,2,3,4,5,6],[10.0,10.0,10.0,10.0,10.0,10.0,10.0]))
// * (3.0,(7,[0,1,2,3,4,5,6],[12.0,12.0,12.0,12.0,12.0,12.0,12.0]))
// * (5.0,(7,[0,1,2,3,4,5,6],[11.0,11.0,11.0,11.0,11.0,11.0,11.0]))
// * (2.0,(7,[0,1,2,3,4,5,6],[2.0,2.0,2.0,2.0,2.0,2.0,2.0]))
// * (5.0,(7,[0,1,2,3,4,5,6],[110.0,110.0,11.0,11.0,11.0,11.0,11.0]))

		// trainData.show();

		JavaPairRDD<LabeledPoint,LabeledPoint> cart = trainData.cartesian(testData);
		cart.saveAsTextFile(args[1]+"/cart");
		
		JavaPairRDD<String,Tuple2<Double,String>> knnMapped =
            //                              input                  				  K       	V
            cart.mapToPair(new PairFunction<Tuple2<LabeledPoint,LabeledPoint>, String, Tuple2<Double,Double>>() {
			// @Override
			public Tuple2<String,Tuple2<Double,Double>> call(Tuple2<LabeledPoint,LabeledPoint> cartRecord) {
				LabeledPoint trainPoint = cartRecord._1;
				LabeledPoint testPoint = cartRecord._2;
				String[] rTokens = trainPoint.split(";"); // TODO need an index of test. that will be the key
				String rRecordID = rTokens[0];
		
				double distance = Util.calcDistance(trainPoint, testPoint);
				String K = rRecordID;//r.recordID; // TODO diff
				Tuple2<Double,Double> V = new Tuple2<Double,Double>(distance, trainPoint.getLabel()); // TODO not string?
				return new Tuple2<String,Tuple2<Double,Double>>(K, V);
			}
		});
		knnMapped.saveAsTextFile(args[1]+"/knnMapped");



		// JavaRDD<String> trainFile = sc.textFile(args[0]);
		// JavaRDD<String> testFile = sc.textFile(args[2]);
		
		// JavaRDD<String> words = textFile.flatMap(new FlatMapFunction<String, String>() {
		// 	public Iterator<String> call(String s) { return Arrays.stream(s.split(" ")).iterator(); }
		// });
		
		// JavaPairRDD<String, Integer> pairs = words.mapToPair(new PairFunction<String, String, Integer>() {
		// 	public Tuple2<String, Integer> call(String s) { return new Tuple2<String, Integer>(s, 1); }
		// });
		
		// JavaPairRDD<String, Integer> counts = pairs.reduceByKey(new Function2<Integer, Integer, Integer>() {
		// 	public Integer call(Integer a, Integer b) { return a + b; }
		// });
		
		// counts.saveAsTextFile(args[1]);

		long endTime = System.nanoTime();
		long totalTime = (endTime - startTime)/1000000;
		// computeAccuracy(args[1], totalTime);
	}
}

File: src/KNN.scala
package com.alwin; //

import org.apache.spark.ml.linalg.DenseVector
import org.apache.spark.sql.SparkSession
import org.apache.spark.sql.DataFrame
import org.apache.spark.sql.Row
// import org.apache.spark.sql.functions
// import org.apache.spark.implicits._
import math._

object KNN {
    def main(args: Array[String]): Unit = { // TODO how to get arg values
        val spark = SparkSession
					.builder
					.appName("DecisionTree")
					.config("spark.master", "local[4]")
					.getOrCreate()
        // import spark.implicits._

        val k = args(3).toInt
		// Load the data stored in LIBSVM format as a DataFrame with dense vectors.
        val trainData = spark.read.format("libsvm").option("vectorType", "dense").load(args(0)) // TODO cache train and/or test?
		val testData = spark.read.format("libsvm").option("vectorType", "dense").load(args(2)) // TODO should be dense and not sparse

        // val testWithIndex = testData.zipWithIndex // ((a,0),(b,1),(c,2)) // TODO need .rdd.zip...?
        // val testIndexKey = testWithIndex.map{case (k,v) => (v,k)} // ((0,a),(1,b),(2,c))
		//Note that some RDDs, such as those returned by groupBy(), do not guarantee order of elements in a partition. The index assigned to each element is therefore not guaranteed, and may even change if the RDD is reevaluated. If a fixed ordering is required to guarantee the same index assignments, you should sort the RDD with sortByKey() or save it to a file.
        
        // val cart = trainData.cartesian(testIndexKey)

        // val knnMapped = cart.map({ cartRecord =>
        //     val trainPoint = cartRecord._1
        //     val testPoint = cartRecord._2
        //     val key = testPoint.getIndex() // TODO getIndex and getValue
        //     val distance = calcDistance(trainPoint, testPoint.getValue());
        //     val value = (distance, trainPoint.getLabel());
        //     (key, value)
		// });
		// knnMapped.saveAsTextFile(args(1)+"/knnMapped");


        println("testData: ", testData)
        testData.collect.foreach(println)
        testData.show

        val nearestKClasses = findKNearestClasses(testData, trainData, k)
        // nearestKClasses.saveAsTextFile(args(1))
        println("nearestKClasses: ", nearestKClasses.mkString(" "))

        spark.stop()
    }

    def calcDistance(xs: Array[Double], ys: Array[Double]) = {
    // def calcDistance(xs: Row, ys: Row) = {
        sqrt((xs zip ys).map { case (x,y) => pow(y - x, 2) }.sum)
    }

    // def findKNearestClasses(testPoints: Array[Array[Double]], trainPoints: Array[Array[Double]], k: Int): Array[Int] = {
    // def findKNearestClasses(testPoints: DataFrame, trainPoints: DataFrame, k: Int): Array[Int] = {
    def findKNearestClasses(testPointsDF: DataFrame, trainPointsDF: DataFrame, k: Int): Array[Double] = {
        val trainPoints = trainPointsDF.select("features").collect.map(row => row.getAs[DenseVector](0).toArray)
        println("inner trainPoints: ")
        trainPoints.foreach(println)
        val trainPointsClasses = trainPointsDF.select("label").collect.map(row => row.getAs[Double](0))
        println("inner trainPointsClasses: ")
        trainPointsClasses.foreach(println)
        // val foo: Nothing = trainPointsClasses
        // println("inner trainPoints string: ", trainPoints.foreach(e=>println(e.getClass)))
        val testPoints = testPointsDF.select("features").collect.map(row => row.getAs[DenseVector](0).toArray)
        println("inner testPoints: ")
        testPoints.map(row => println(row.toArray.mkString(" ")))
        println("")

        // val trainPoints = trainPointsDF.map(row=>row.getSeq[Double](1)).collect
        // println("trainPoints: ", trainPoints.foreach(println))
        // val testPoints = testPointsDF.map(row=>row.getAs[Array[DoubleType]](1).toArray).collect

        // val trainPoints = trainPointsDF.rdd.map(row=>row.getAs[Vector[Double]](1).toArray).collect
        // println("trainPoints: ", trainPoints)
        // val testPoints = testPointsDF.rdd.map(row=>row.getAs[Vector[Double]](1).toArray).collect

        // val trainPoints = trainPointsDF.rdd.map(row=>row.getAs[Array[Double]](1)).collect
        // println("trainPoints: ", trainPoints)
        // val testPoints = testPointsDF.rdd.map(row=>row.getAs[Array[Double]](1)).collect

        // val trainPoints = trainPointsDF.rdd.map(row=>row.getAs[Double](0) -> row.getAs[Vector](1).toArray).toDF("class","array")
        // println("trainPoints: ", trainPoints)
        // val testPoints = testPointsDF.rdd.map(row=>row.getAs[Double](0) -> row.getAs[Vector](1).toArray).toDF("class","array")

        // val trainPoints = trainPointsDF.rdd.map(row=>row.getDouble(1)).collect
        // println("trainPoints: ", trainPoints)
        // val testPoints = testPointsDF.rdd.map(row=>row.getDouble(1)).collect

        // val trainPoints = trainPointsDF.rdd.map(row=>row.getDouble(1).toArray).collect //trainPointsDF.rdd.map(row=>row.getDouble(0)).collect
        // val testPoints = testPointsDF.rdd.map(row=>row.getDouble(1).toArray).collect
        
        
        // return new Array[Int](1)

        // val trainPoints = trainPointsDF.select("features").collect.map(row => row.getAs[DenseVector](0).toArray)

        testPoints.map { testInstance =>
            val distances = 
                (trainPoints zip trainPointsClasses).map { case (trainInstance, c) =>
                    c->println("trainInstance: ", trainInstance.mkString(" "))
                    c -> calcDistance(testInstance, trainInstance)
                }
            println("distances: ")
            distances.foreach(println)
            val classes = distances.sortBy(_._2).take(k).map(_._1)
            println("classes: ")
            classes.foreach(println)
            // val foo: Nothing = classes
            val classCounts = classes.groupBy(identity).mapValues(_.size)
            println("classCounts: ", classCounts)
            classCounts.maxBy(_._2)._1
        }

        // testPoints.map { testInstance =>
        //     val distances = 
        //         trainPoints.zipWithIndex.map { case (trainInstance, c) =>
        //             c->println("trainInstance: ", trainInstance.mkString(" "))
        //             c -> calcDistance(testInstance, trainInstance)
        //         }
        //     // val foo: Nothing = distances
        //     println("distances: ")
        //     distances.foreach(println)
        //     val classes = distances.sortBy(_._2).take(k).map(_._1)
        //     println("classes: ")
        //     classes.foreach(println)
        //     // val foo: Nothing = classes
        //     val classCounts = classes.groupBy(identity).mapValues(_.size)
        //     println("classCounts: ", classCounts)
        //     classCounts.maxBy(_._2)._1
        // }
    }


}